tests/src/helpers.c

```diff
@@ -244,3 +244,16 @@ void mbedtls_param_failed( const char *failure_condition,
     }
 }
 #endif /* MBEDTLS_CHECK_PARAMS */
+
+#if defined(MBEDTLS_TEST_HOOKS)
+void mbedtls_test_err_add_check( int high, int low,
+                                 const char *file, int line )
+{
+    if ( high > -0x1000 || low < -0x007F )
+    {
+        mbedtls_fprintf( stderr, "\nIncorrect error code addition at %s:%d\n",
+                                file, line );
+        mbedtls_exit( 1 );
```

Comment:
This is very unfriendly. The test case won't be reported as failed, and subsequent test cases won't run.

The right thing to do is to mark the test case as failed. At least, I think that's a good approximation: the error location will be at the place where `MBEDTLS_ERR_ADD` is called rather than in the test suite, so the programmer might end up needing a debugger to figure out how the problematic function was called, but we can't easily do better.

Marking the test as failed is annoyingly difficult because `test_fail` is defined in `helpers.function` and therefore can't be called from `tests/src/helpers.c`. I guess it's time to move `test_fail` and friends to a new file in `tests/src`, or maybe `tests/src/helpers.c` itself. This is a natural follow-up to https://github.com/ARMmbed/mbedtls/pull/3387 that we just haven't got around to yet. This is not the first time we've hit this limitation with `test_fail`, for example it bit me recently in https://github.com/ARMmbed/mbedtls/pull/3912#discussion_r555668044.

Chris, do you have the time to do the move of `test_fail` and friends (maybe chat with Ronald or me first to agree on a scope), then rebase `mbedtls_test_err_add_check` on top of that?

Reply:
Yes I agree this is not how I wanted to do it and this was simply a temporary workaround of that issue. I was going to use `TEST_HELPER_ASSERT` as that is in scope of `helpers.c` however decided against it as it also doesn't report the correct location of the failure as well as just exiting instead of failing the test.

I will take a look at moving `test_fail` now as I'd be happy to do the work.

Reply:
#4028 and #4054 have been created to address this issue.

Reply:
#4054 has been merged and this PR has been rebased on top to now include these changes.

I have also added a commit to use `mbedtls_test_fail` as opposed to manually printing and exiting as `mbedtls_test_fail` is now available.